project/src/backend/sdl3/SDLSystem.cpp

```diff
@@ -0,0 +1,1003 @@
+				#else
+				dpi = pixelDensity * contentScale * 96.0f;
+				#endif
+			#endif
+			hl_dyn_setf (display, id_dpi, dpi);
+
+			SDL_DisplayOrientation orientation = SDL_GetCurrentDisplayOrientation(id);
+			hl_dyn_seti (display, id_orientation, &hlt_i32, orientation);
+
+			DisplayMode mode;
+
+			const SDL_DisplayMode *displayMode = SDL_GetDesktopDisplayMode (id);
+
+			mode.height = displayMode->h;
+
+			switch (displayMode->format) {
+
+				case SDL_PIXELFORMAT_ARGB8888:
+
+					mode.pixelFormat = ARGB32;
+					break;
+
+				case SDL_PIXELFORMAT_BGRA8888:
+				case SDL_PIXELFORMAT_BGRX8888:
+
+					mode.pixelFormat = BGRA32;
+					break;
+
+				default:
+
+					mode.pixelFormat = RGBA32;
+
+			}
+
+			mode.refreshRate = displayMode->refresh_rate;
+			mode.width = displayMode->w;
+
+			vdynamic* _displayMode = (vdynamic*)hl_alloc_dynobj ();
+			hl_dyn_seti (_displayMode, id_height, &hlt_i32, mode.height);
+			hl_dyn_seti (_displayMode, id_pixelFormat, &hlt_i32, mode.pixelFormat);
+			hl_dyn_seti (_displayMode, id_refreshRate, &hlt_i32, mode.refreshRate);
+			hl_dyn_seti (_displayMode, id_width, &hlt_i32, mode.width);
+			hl_dyn_setp (display, id_currentMode, &hlt_dynobj, _displayMode);
+
+			SDL_DisplayMode **displayModes = SDL_GetFullscreenDisplayModes (id, NULL);
+			int numDisplayModes = sizeof(displayModes) / sizeof(displayModes[0]);
+
+			hl_varray* supportedModes = (hl_varray*)hl_alloc_array (&hlt_dynobj, numDisplayModes);
+			vdynamic** supportedModesData = hl_aptr (supportedModes, vdynamic*);
+
+			for (int i = 0; i < numDisplayModes; i++) {
+
+				displayModes[i];
+
+				mode.height = displayMode->h;
+
+				switch (displayMode->format) {
+
+					case SDL_PIXELFORMAT_ARGB8888:
+
+						mode.pixelFormat = ARGB32;
+						break;
+
+					case SDL_PIXELFORMAT_BGRA8888:
+					case SDL_PIXELFORMAT_BGRX8888:
+
+						mode.pixelFormat = BGRA32;
+						break;
+
+					default:
+
+						mode.pixelFormat = RGBA32;
+
+				}
+
+				mode.refreshRate = displayMode->refresh_rate;
+				mode.width = displayMode->w;
+
+				vdynamic* _displayMode = (vdynamic*)hl_alloc_dynobj ();
+				hl_dyn_seti (_displayMode, id_height, &hlt_i32, mode.height);
+				hl_dyn_seti (_displayMode, id_pixelFormat, &hlt_i32, mode.pixelFormat);
+				hl_dyn_seti (_displayMode, id_refreshRate, &hlt_i32, mode.refreshRate);
+				hl_dyn_seti (_displayMode, id_width, &hlt_i32, mode.width);
+
+				*supportedModesData++ = _displayMode;
+
+			}
+
+			hl_dyn_setp (display, id_supportedModes, &hlt_array, supportedModes);
+			return display;
+
+		}
+
+	}
+
+
+	int System::GetNumDisplays () {
+
+		return (sizeof(SDL_GetDisplays(NULL)) / sizeof(SDL_GetDisplays(NULL)[0]));
+
+	}
+
+
+	double System::GetTimer () {
+
+		return SDL_GetTicks ();
+
+	}
+
+
+	bool System::SetAllowScreenTimeout (bool allow) {
+
+		if (allow) {
+
+			SDL_EnableScreenSaver ();
+
+		} else {
+
+			SDL_DisableScreenSaver ();
+
+		}
+
+		return allow;
+
+	}
+
+
+	FILE* FILE_HANDLE::getFile () {
+		return (FILE*)handle;
```

Comment:
The SDL_IOStream is now an opaque struct, so this cast no longer safe and causes segmentation faults on Linux when the FILE* is passed to png functions. Accessing it via the property solves this problem:

```suggestion
		SDL_PropertiesID properties = SDL_GetIOProperties((SDL_IOStream*)handle);
		return (FILE*)SDL_GetPointerProperty(properties, SDL_PROP_IOSTREAM_STDIO_FILE_POINTER, NULL);
```
> The internal FILE * is available through a standard SDL_IOStream property, for streams made through SDL_IOFromFile() that use stdio behind the scenes; apps use this pointer at their own risk and should make sure that SDL and the app are using the same C runtime.

https://wiki.libsdl.org/SDL3/README/migration#sdl_rwopsh

`SDL_IOFromFile` sets this property, see also:
https://wiki.libsdl.org/SDL3/SDL_IOFromFile#remarks

Reply:
> We may need separate logic for android and possibly windows

Compare to the sdl2 implementation:
https://github.com/Fancy2209/lime/blob/d74b0cbcf972b3287c02ef5a357543275e13a7b7/project/src/backend/sdl/SDLSystem.cpp#L602-L641